R/params.R

```diff
@@ -0,0 +1,329 @@
+  ## A value might have multiple classes. Try: class(Sys.time())
+  ## Try to find first class listed with a named control.
+  for (c in param$class) {
+    control <- params_controls[[c]]
+    if (!is.null(control)) {
+      return(control(param))
+    }
+  }
+  return(NULL)
+}
+
+params_configurable <- function(param) {
+  (!is.null(params_get_control(param))) && ( length(param$value) == 1 || "data.frame" %in% param$class )
+}
+
+#' Run a shiny application asking for parameter configuration for the given document.
+#'
+#' @param file Path to the R Markdown document with configurable parameters.
+#' @param input_lines Content of the R Markdown document. If \code{NULL}, the contents of \code{file} will be read.
+#' @param params A named list of optional parameter overrides used in place of the document defaults.
+#' @param shiny_args Additional arguments to \code{\link[shiny:runApp]{runApp}}.
+#'
+#' @return named list with overridden parameter names and value.
+#' 
+#' @export
+knit_params_ask <- function(file = "index.Rmd",
+                            input_lines = NULL,
+                            params = NULL,
+                            shiny_args = NULL,
+                            encoding = getOption("encoding")) {
+  if (packageVersion("knitr") < "1.10.17") {
+    stop("knitr >= 1.10.17 required to use rmarkdown::knit_params_ask")
+  }
+
+  if (is.null(input_lines)) {
+    input_lines <- read_lines_utf8(file, encoding)
+  }
+  
+  knit_params <- mark_utf8(knitr::knit_params(input_lines))
+  configurable <- Filter(params_configurable, knit_params)
+  unconfigurable <- Filter(Negate(params_configurable), knit_params)
+
+  ## Input validation on params (checks shared with render)
+  if (!is.null(params)) {
+    ## Must be a named list
+    if (!is.list(params) || (length(names(params)) != length(params))) {
+      stop("knit_params_ask params argument must be a named list")
+    }
+
+    ## Verify that all passed parameters are also in the yaml
+    invalid_params <- setdiff(names(params), names(knit_params))
+    if (length(invalid_params) > 0) {
+      stop("knit_params_ask params not declared in YAML: ",
+           paste(invalid_params, sep = ", "))
+    }
+  }
+  
+  ## Return the current value for a given parameter, which is either `params`
+  ## when overridden by our caller or `param$value` otherwise.
+  current_value <- function(param) {
+    if (!is.null(params)) {
+      override <- params[[param$name]]
+      if (!is.null(override)) {
+        return(override)
+      }
+    }
+    return(param$value)
+  }
+
+  ## The default values given to the UI. Lets us detect if the user has
+  ## specified an override value or is letting it float to whatever default is
+  ## generated by the YAML (most useful for R expressions).
+  ##
+  ## Only items in `configurable` will be present here.
+  uidefaults <- list()
+
+  server <- function(input, output) {
+    param.ui <- function(param) {
+      inputId <- paste0(param$name)
```

Comment:
I had some problems with labels in early versions of this, and believed that this helped. But I can't see why any longer. Removing.